MLflow 3.0 docs (#15078)
mlflow/mlflow: docs/docs/mlflow-3/index.mdx

## What is MLflow 3.0?

MLflow 3.0 delivers best-in-class experiment tracking, observability, and performance evaluation for machine learning models, AI applications, and generative AI agents!
MLflow 3.0 introduces a variety of exciting improvements, including:
- Centrally track and analyze the performance of your models and generative AI applications **across all environments**, from interactive queries in a development notebook through production batch or real-time serving deployments.
- Gain insights into model performance with **comprehensive metrics and lineage tracking**, enabling better decision-making for production readiness.
- Streamline the handling of model artifacts and evaluation results with a **cohesive structure** that improves observability and traceability.

Model Tracking helps you create and evaluate different model/agent configurations in your experiments, <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink> helps you find the best ones for production usage, and tracing helps enhance observability in your model or GenAI applications.

> ## Model Tracking with LoggedModel

Reviewer: ```suggestion
## Enhanced Model Tracking
```

> In MLflow 3.0, we introduce a refined architecture along with revamped APIs and UI, tailored to enhance deep learning and generative AI workflows. 

Reviewer: @serena-ruan Can we switch the order of deep learning and generative AI throughout the sentences in these docs? The emphasis of MLflow 3.0 should be GenAI.

Can we also switch the order of the deep learning and GenAI sub-tabs?

In deep learning, training often generates multiple model checkpoints, where the best candidates are further evaluated before production deployment. 
With GenAI agents, there are multiple rounds of offline evaluation via batch jobs and interactive queries from human beta testers. 

> We are introducing a new first-class object, the Logged Model entity, into MLflow Tracking to streamline these processes. As your deep learning jobs create and evaluate models, 

Reviewer: Can we add a hyperlink to `Logged Model` here?

---
sidebar_label: MLflow 3.0
sidebar_position: 1
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import { APILink } from "@site/src/components/APILink";
import { CardGroup, PageCard } from "@site/src/components/Card";

> # MLflow 3.0

Reviewer: ```suggestion
# MLflow 3.0 (Preview)
```

### Quickstart

This quickstart example demonstrates how to use the LoggedModel lineage feature in MLflow 3.0 with a custom Python model, showcasing its integration with runs and traces.

```python
import mlflow

class MyModel(mlflow.pyfunc.PythonModel):
    @mlflow.trace
    def predict(self, model_input: list[str]):
        return ",".join(model_input)

with mlflow.start_run():
    model_info = mlflow.pyfunc.log_model(
        python_model=MyModel(), name="my_model", input_example=["a", "b", "c"]
    )

print(f"LoggedModel model id: {model_info.model_id}")

pyfunc_model = mlflow.pyfunc.load_model(model_info.model_uri)
pyfunc_model.predict(["x", "y", "z"])
traces = mlflow.search_traces(model_id=model_info.model_id)
print(traces)

#                          request_id                                              trace   timestamp_ms  ...                                              spans                                               tags assessments
# 0  af5da8c78a6f49c793f57ebb19dcb6c7  Trace(request_id=af5da8c78a6f49c793f57ebb19dcb...  1742781867233  ...  [{'name': 'predict', 'context': {'span_id': '0...  {'mlflow.artifactLocation': 'file:///Users/ser...          []
```

> Navigate to the run's page to view the LoggedModel:

Reviewer: The metrics and params are blank in this example, so the screenshot doesn't look super compelling. Can we include an example that has metrics and params?

---
sidebar_label: MLflow 3.0
sidebar_position: 1
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import { APILink } from "@site/src/components/APILink";
import { CardGroup, PageCard } from "@site/src/components/Card";

# MLflow 3.0 (Preview)

Discover the next generation of MLflow, designed to streamline your AI experimentation and accelerate your journey from idea to production. MLflow 3.0 brings cutting-edge support for GenAI workflows, enabling seamless integration of generative AI models into your projects.

## What is MLflow 3.0?

> MLflow 3.0 delivers best-in-class experiment tracking, observability, and performance evaluation for machine learning models, AI applications, and generative AI agents!
> MLflow 3.0 introduces a variety of exciting improvements, including:
> - Centrally track and analyze the performance of your models and generative AI applications **across all environments**, from interactive queries in a development notebook through production batch or real-time serving deployments.
> - Gain insights into model performance with **comprehensive metrics and lineage tracking**, enabling better decision-making for production readiness.
> - Streamline the handling of model artifacts and evaluation results with a **cohesive structure** that improves observability and traceability.

Reviewer: ```suggestion
MLflow 3.0 delivers best-in-class experiment tracking, observability, and performance evaluation for machine learning models, AI applications, and generative AI agents!
With MLflow 3.0, it's now easier than ever to:
- Centrally track and analyze the performance of your models, agents, and generative AI applications **across all environments**, from interactive queries in a development notebook through production batch or real-time serving deployments.
- Select the best models, agents, and generative AI applications for production with an **enhanced performance comparison experience** powered by MLflow's tracing and evaluation capabilities.
```

The third bullet point isn't particularly clear. As we get closer to the stable release, we can add a third bullet point about feedback tracking / human labeling for generative AI.

> Model Tracking helps you create and evaluate different model/agent configurations in your experiments, <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink> helps you find the best ones for production usage, and tracing helps enhance observability in your model or GenAI applications.

Reviewer: ```suggestion
MLflow 3.0's enhanced model tracking helps you manage and evaluate different model, agent, and generative AI application configurations in your experiments. Improved comparison workflows with the new <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink> enable you to quickly identify the best candidates for production, and MLflow tracing provides rich observability everywhere that your models, agents, and generative AI applications are deployed.
```

## Enhanced Model Tracking

In MLflow 3.0, we introduce a refined architecture along with revamped APIs and UI, tailored to enhance generative AI and deep learning workflows. 
With GenAI agents, there are multiple rounds of offline evaluation via batch jobs and interactive queries from human beta testers.
In deep learning, training often generates multiple model checkpoints, where the best candidates are further evaluated before production deployment. 

We are introducing a new first-class object, the <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink> entity, into MLflow Tracking to streamline these processes. As you define and evaluate your GenAI agents in code, 
or your deep learning jobs create and evaluate models, they will be automatically stored as MLflow Logged Models in your MLflow Experiment. 

> Each GenAI agent consolidates its evaluation metrics from both offline jobs and online interactions if you use managed MLflow with Databricks. 
> This feature removes the need to query Runs for traces from different environments (batch jobs, different notebook sessions etc.) to gain a full understanding of the agent’s performance.
> Similarly, every deep learning checkpoint is stored as a <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink> with direct links to its own metrics. There is no longer any need to carefully log evaluation metrics for each checkpoint

Reviewer: ```suggestion
Each GenAI agent consolidates its evaluation metrics from both offline jobs and online interactions.
Similarly, every deep learning checkpoint is stored as a <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink> with direct links to its own metrics. There is no longer any need to carefully log evaluation metrics for each checkpoint
```

> Each GenAI agent consolidates its evaluation metrics from both offline jobs and online interactions if you use managed MLflow with Databricks. 
> This feature removes the need to query Runs for traces from different environments (batch jobs, different notebook sessions etc.) to gain a full understanding of the agent’s performance.
> Similarly, every deep learning checkpoint is stored as a <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink> with direct links to its own metrics. There is no longer any need to carefully log evaluation metrics for each checkpoint
> at its epoch all within the same Run, or to later query all metrics at a certain epoch just to retrieve metrics for one checkpoint. 

Reviewer: ```suggestion
Each GenAI agent consolidates its evaluation metrics from both offline jobs and online interactions.
Similarly, every deep learning checkpoint is stored as a <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink> with direct links to its own metrics.
```

> Specifically, you will be able to:

Reviewer: ```suggestion
With <APILink fn="mlflow.entities.LoggedModel">LoggedModel</APILink>'s, you can:
```

> Specifically, you will be able to:
> - View Logged Models alongside key information, including code and weights as artifacts and configurable parameters that affect model behavior
> - Holistically track all performance metrics related to your logged models with links to training and evaluation jobs
> - Search for and rank existing models based on a set of certain criteria, such as performance metrics on training or evaluation datasets

Reviewer: ```suggestion
```

Runs already offer these capabilities, so users will wonder why LoggedModels are special. Let's remove this section.

### Quickstart

> This quickstart example demonstrates how to use the LoggedModel lineage feature in MLflow 3.0 with openai sdk, showcasing its integration with runs and traces.

Reviewer: Can we explain that this example creates a generative AI application with prompt engineering and then evaluates it?

> Prerequisite: ``pip install mlflow==3.0.0.rc0 openai``

Reviewer: ```suggestion
Prerequisite: ``pip install mlflow>=3.0.0.rc0 openai``
```

```python
from openai import OpenAI

import mlflow
from mlflow.metrics.genai import answer_correctness, answer_similarity, faithfulness

# turn on autologging for automatic tracing
mlflow.openai.autolog()

> client = OpenAI()

Reviewer: Can we explain that we're defining a simple generative AI application that answers questions about MLflow using prompt engineering? Can we also include a default prompt?

Author: This example doesn't include prompt registry, the detailed genai one includes it. Do we want to include here?

mlflow_ground_truth = (
    "MLflow is an open-source platform for managing "
    "the end-to-end machine learning (ML) lifecycle. It was developed by Databricks, "
    "a company that specializes in big data and machine learning solutions. MLflow is "
    "designed to address the challenges that data scientists and machine learning "
    "engineers face when developing, training, and deploying machine learning models."
)
# Define evaluation metrics
metrics = {
    "answer_similarity": answer_similarity(model="openai:/gpt-4o"),
    "answer_correctness": answer_correctness(model="openai:/gpt-4o"),
    "faithfulness": faithfulness(model="openai:/gpt-4o"),
}
question = "What is MLflow?"

> with mlflow.start_run():

Reviewer: Why do I need a run here?

> with mlflow.start_run():

Reviewer: Can we explain why we're starting a run in the comments? E.g. because we're running an evaluation?

            model="gpt-4o-mini",
            temperature=0.1,
            max_tokens=2000,
        )
        .choices[0]
        .message.content
    )

    # Calculate metrics based on the input, response and ground truth
    # The evaluation metrics are callables that can be invoked directly
    answer_similarity_score = metrics["answer_similarity"](
        predictions=response, inputs=question, targets=mlflow_ground_truth
    ).scores[0]
    answer_correctness_score = metrics["answer_correctness"](
        predictions=response, inputs=question, targets=mlflow_ground_truth
    ).scores[0]
    faithfulness_score = metrics["faithfulness"](
        predictions=response, inputs=question, context=mlflow_ground_truth
    ).scores[0]

    # Fetch the LoggedModel that's automatically created during autologging
    logged_model = mlflow.last_logged_model()
    # Log metrics and pass model_id to link the metrics
    mlflow.log_metrics(
        {
            "answer_similarity": answer_similarity_score,
            "answer_correctness": answer_correctness_score,
            "faithfulness": faithfulness_score,
        },
        model_id=logged_model.model_id,

>     )

Reviewer: Is linking by `model_id` necessary? I thought MLflow would automatically figure this out based on the logged model

Author: It's necessary right now because the model is not logged inside a evaluation run, it was created by autologging during invocation

print(f"LoggedModel model id: {logged_model.model_id}")
# LoggedModel model id: a208e70b-e80b-4f8e-b210-20737faadd20

traces = mlflow.search_traces(model_id=logged_model.model_id)
print(traces)
#                          request_id                                              trace  ...                                               tags assessments
# 0  5882df1240cf4dbf845fdc9fa26c4168  Trace(request_id=5882df1240cf4dbf845fdc9fa26c4...  ...  {'mlflow.artifactLocation': 'file:///Users/ser...          []
# [1 rows x 11 columns]
```

> Navigate to the run's page to view the LoggedModel:

Reviewer: We should show the model's page first and explain that it shows metrics from all evaluation jobs. At the *end*, we should say that users can navigate navigate to an individual evaluation job by clicking the link from the model page & we should show the runs page *last*

---
sidebar_label: MLflow 3.0
sidebar_position: 1
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import Platform from "@site/src/content/platform.mdx"

# MLflow 3.0

Discover the next generation of MLflow, designed to streamline your AI experimentation and accelerate your journey from idea to production. MLflow 3.0 brings cutting-edge support for GenAI workflows, enabling seamless integration of generative AI models into your projects.

## What is MLflow 3.0?

The best AI experimentation platform for building high quality models and GenAI agents! MLflow 3.0 introduces a variety of exciting improvements to MLflow Tracking and Tracing, including:

> - First class model support in MLflow tracking with **LoggedModel**

Reviewer 2: Link `LoggedModel` to API docs?

Reviewer 3: +1

- **Model-centric** information architecture for params and metrics in MLflow tracking
- Enhanced **tracing** experience with lineage tracking for LoggedModels

Model Tracking helps you create and evaluate different model/agent configurations in your experiments, LoggedModel helps you find the best ones for production usage, and tracing helps enhance observability in your model or GenAI applications.

> 

Reviewer 2: Can we add a pseudo table of contents so that users can drive to the sections "MLflow 3 for GenAI" "MLflow 3 for Deep Learning" as an indicator that "this isn't just for GenAI". Perhaps consolidate the GenAI examples together and have the DL example at the bottom.

Author: <img width="290" alt="image" src="https://github.com/user-attachments/assets/337fd637-496b-4755-90b7-eeeee5539732" />
The ToC on the right is quite straightforward 🤔 Do we need an extra list here? Maybe if we move them into separate pages later and adding a table of contents here makes more sense?

## Model Tracking with LoggedModel

In MLflow 3.0, we introduce a new model-centric design, together with revamped APIs (UI will come later), tailored to enhance deep learning and generative AI workflows. 
In deep learning, training often generates multiple model checkpoints, where the best candidates are further evaluated before production deployment. 
With GenAI agents, there are multiple rounds of offline evaluation via batch jobs and interactive queries from human beta testers. 

We are introducing a new first-class object, the Logged Model entity, into MLflow Tracking to streamline these processes. As your deep learning jobs create and evaluate models, 
or you define and evaluate your GenAI agents in code, they will be automatically stored as MLflow Logged Models in your MLflow Experiment. 

Every deep learning checkpoint is stored as a Logged Model with direct links to its own metrics. There is no longer any need to carefully log evaluation metrics for each checkpoint
at its epoch all within the same Run, or to later query all metrics at a certain epoch just to retrieve metrics for one checkpoint. 

> Similarly, each GenAI agent consolidates its evaluation metrics from offline jobs, and online interactions if you use managed MLflow with Databricks. 

Reviewer 2: ```suggestion
Similarly, each GenAI agent consolidates its evaluation metrics from both offline jobs and online interactions if you use managed MLflow with Databricks. 
```

> This feature removes the need to query Runs for traces from different environments (batch jobs, separate inference tables etc.) to gain a full understanding of the agent’s performance.

Reviewer 3: Can we remove `inference tables` because they don't exist?

Specifically, you will be able to:
- View Logged Models alongside key information, including code and weights as artifacts and configurable parameters that affect model behavior
- Holistically track all performance metrics related to your logged models with links to training and evaluation jobs
- Search for and rank existing models based on a set of certain criteria, such as performance metrics on training or evaluation datasets

Simple example:

```python
import mlflow

> class MyModel(mlflow.pyfunc.PythonModel):

Reviewer 2: Since all of our recommended paths for model logging for GenAI are going to be focused on models from code, should we update this example to highlight that? It would good to reinforce this paradigm in all examples, even though it is slightly more complex.

Author: Since mlflow-3 is not only for GenAI, and this example is super simple maybe we can keep it as it is? I'll update the GenAI one to use models from code though!

Reviewer: @BenWilson2 @serena-ruan instead of the "hello world" example, I'd prefer if we show a compelling *screenshot* of a GenAI example containing LoggedModel params and metrics. I don't think there's a ton of value in showing "hello world" examples since we have full GenAI / DL examples on the next page, since they aren't indicative of real workflows and can make the product appear "out of touch"

```

### GenAI Example

This example demonstrates how to use MLflow to log and evaluate a generative AI agent.
It showcases how to register prompts, log models, and assess their performance using evaluation datasets.
The example also highlights the ability to track interactive traces and link them to the logged model for better observability.

Prerequisite: ``pip install langchain langchain-community langchain-openai``

```python
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI

import mlflow

# register a prompt so we can link it when logging the model
system_prompt = mlflow.register_prompt(
    name="system_prompt",
    template="You are a chatbot that can answer questions about IT.",
    commit_message="Initial version of system prompt",
)

chat_model = ChatOpenAI(
    model="gpt-4o-mini", temperature=0.1, max_completion_tokens=2000
)
chat_prompt = ChatPromptTemplate.from_messages(
    [("system", system_prompt.to_single_brace_format()), ("user", "{messages}")]
)

> chain = chat_prompt | chat_model

Reviewer 2: Can we change this to models from code approach?

        ],
        # This is needed since answer_correctness looks for 'inputs' field
        evaluator_config={"col_mapping": {"inputs": "messages"}},
    )

print(result.tables["eval_results_table"])
#                                         messages                                  expected_response  ... answer_correctness/v1/score                answer_correctness/v1/justification
# 0  What is MLflow Tracking and how does it work?  MLflow Tracking is a key component of the MLfl...  ...                           5  The output accurately describes MLflow Trackin...
# 1                         What is Unity Catalog?  Unity Catalog is a feature in Databricks that ...  ...                           5  The output accurately describes Unity Catalog ...
# 2        What are user-defined functions (UDFs)?  User-defined functions (UDFs) in the context o...  ...                           5  The output accurately describes user-defined f...
```

## Migration Guide

MLflow 3.0 introduces some key API changes while also removes some outdated features. This guide will help you transition smoothly to the latest version.

### Key changes

> - `mlflow.<flavor>.log_model` API usage: `artifact_path` parameter is deprecated, use **`name`** instead

Reviewer 2: Can we show a before / after comparison with code here to make it more clear to users what is changing?

> - Model artifacts storage location change: In MLflow 2.x, [model artifacts](../model/#storage-format) are stored as run artifacts. Since MLflow 3.0, those artifacts will be stored into models artifacts location.

Reviewer 2: Similar here. Can we show the paths differences so that users understand the change in storage location?

Reviewer 3: Let's also mention that this will impact `list_artifacts`.

Author: I'll add a screenshot later when UI is merged :)

    mlflow.log_metrics(
        metrics=result.metrics,
        dataset=eval_dataset,
        # Specify the ID of the agent logged above
        model_id=logged_model.model_id,
    )

print(result.tables["eval_results_table"])
#                                         messages                                  expected_response  ... answer_correctness/v1/score                answer_correctness/v1/justification
# 0  What is MLflow Tracking and how does it work?  MLflow Tracking is a key component of the MLfl...  ...                           5  The output accurately describes MLflow Trackin...
# 1                         What is Unity Catalog?  Unity Catalog is a feature in Databricks that ...  ...                           5  The output accurately describes Unity Catalog ...
# 2        What are user-defined functions (UDFs)?  User-defined functions (UDFs) in the context o...  ...                           5  The output accurately describes user-defined f...
```

## Migration Guide

MLflow 3.0 introduces some key API changes while also removes some outdated features. This guide will help you transition smoothly to the latest version.

### Key API changes

> - `mlflow.<flavor>.log_model`: `artifact_path` parameter is deprecated, use **`name`** instead

Author: @harupy Is there any other outstanding API changes we should mention here?

Reviewer 3: Not a API change, but model won't show up as run artifacts on MLflow UI.

Author: Databricks UI shows the artifacts (with a reminder to route users to models page), will we do the same for OSS UI?

Reviewer 3: I think that will also be exported to the mlflow UI.

print(f"Worst model: {worst_checkpoint}")
print(worst_checkpoint.metrics)

# Worst model: <LoggedModel: artifact_location='file:///Users/serena.ruan/Documents/repos/mlflow/mlruns/0/models/da03f02f-1352-456a-97dc-3cf4f1f83fae/artifacts', creation_timestamp=1742784476515, experiment_id='0', last_updated_timestamp=1742784483429, metrics=[<Metric: dataset_digest='1f1c13b5', dataset_name='train', key='accuracy', model_id='da03f02f-1352-456a-97dc-3cf4f1f83fae', run_id='92b65129f25a412f9455e637a3d95a2a', step=0, timestamp=1742784483447, value=0.3333333333333333>], model_id='da03f02f-1352-456a-97dc-3cf4f1f83fae', model_type=None, model_uri='models:/da03f02f-1352-456a-97dc-3cf4f1f83fae', name='torch-iris-0', params={'activation': 'ReLU',
# 'criterion': 'CrossEntropyLoss',
# 'n_layers': '3',
# 'optimizer': 'Adam'}, source_run_id='92b65129f25a412f9455e637a3d95a2a', status=<LoggedModelStatus.READY: 'READY'>, status_message=None, tags={}>
# [<Metric: dataset_digest='1f1c13b5', dataset_name='train', key='accuracy', model_id='da03f02f-1352-456a-97dc-3cf4f1f83fae', run_id='92b65129f25a412f9455e637a3d95a2a', step=0, timestamp=1742784483447, value=0.3333333333333333>]
```

> ### GenAI Example

Reviewer 3: Can we create separate pages for each example, and link them in `index.mdx`?

Author: I would prefer to do so later when we have UI to explain the lineage, right now it's quite short to fit in a separate page 🤔

Reviewer 3: Makes sense :)

        ],
    }
)

# Start a run to represent the evaluation job
with mlflow.start_run() as evaluation_run:
    eval_dataset = mlflow.data.from_pandas(
        df=eval_df, name="eval_dataset", targets="expected_response"
    )
    # Run the evaluation based on extra metrics
    result = mlflow.evaluate(
        model=f"models:/{logged_model.model_id}",
        data=eval_dataset,
        # targets="expected_response",
        model_type="question-answering",
        extra_metrics=[
            mlflow.metrics.latency(),
            mlflow.metrics.genai.answer_correctness("openai:/gpt-4o-mini"),
        ],
        # This is needed since answer_correctness looks for 'inputs' field
        evaluator_config={"col_mapping": {"inputs": "messages"}},
    )
    # Log evaluation metrics and associate with agent
    mlflow.log_metrics(
        metrics=result.metrics,
        dataset=eval_dataset,
        # Specify the ID of the agent logged above
        model_id=logged_model.model_id,

>     )

Reviewer 3: Do we need this? Does `mlflow.evaluate` automatically link the metrics to the model?

# 0  What is MLflow Tracking and how does it work?  MLflow Tracking is a key component of the MLfl...  ...                           5  The output accurately describes MLflow Trackin...
# 1                         What is Unity Catalog?  Unity Catalog is a feature in Databricks that ...  ...                           5  The output accurately describes Unity Catalog ...
# 2        What are user-defined functions (UDFs)?  User-defined functions (UDFs) in the context o...  ...                           5  The output accurately describes user-defined f...
```

## Migration Guide

MLflow 3.0 introduces some key API changes while also removes some outdated features. This guide will help you transition smoothly to the latest version.

### Key changes

- `mlflow.<flavor>.log_model` API usage: `artifact_path` parameter is deprecated, use **`name`** instead
- Model artifacts storage location change: In MLflow 2.x, [model artifacts](../model/#storage-format) are stored as run artifacts. Since MLflow 3.0, those artifacts will be stored into models artifacts location.

### Removed Features

- MLflow Recipes
- Flavors: the following model flavors are no longer supported
    - fastai

>     - h2o
> 	- mleap

Reviewer 3: ```suggestion
    - h2o
    - mleap
```

### Key changes

- `mlflow.<flavor>.log_model` API usage: `artifact_path` parameter is deprecated, use **`name`** instead
<Tabs>
    <TabItem label="MLflow 2.x" value="mlflow_2" default>
        ```python
        with mlflow.start_run():
            mlflow.pyfunc.log_model(artifact_path="model", python_model=python_model, ...)
        ```
    </TabItem>
    <TabItem label="MLflow 3.0" value="mlflow_3">
        Pass `name` when logging a model. This allows you to later search for LoggedModels using this name.
        ```python
        with mlflow.start_run():
            mlflow.pyfunc.log_model(name="python_model", python_model=python_model, ...)
        ```
    </TabItem>
</Tabs>

- Model artifacts storage location change: In MLflow 2.x, [model artifacts](../model/#storage-format) are stored as run artifacts. 
Since MLflow 3.0, those artifacts will be stored into models artifacts location. Note: this impacts the behavior of <APILink fn="mlflow.client.MlflowClient.list_artifacts">``list_artifacts``</APILink> API.

### Removed Features

- MLflow Recipes
- Flavors: the following model flavors are no longer supported
    - fastai
    - h2o
    - mleap

> - AI gateway APIs: use deployments APIs instead

Reviewer: ```suggestion
- AI gateway client APIs: use deployments APIs instead
```

## What is MLflow 3.0?

> The best AI experimentation platform for building high quality models and GenAI agents! MLflow 3.0 introduces a variety of exciting improvements to MLflow Tracking and Tracing, including:

Reviewer: ```suggestion
MLflow 3.0 delivers best-in-class experiment tracking, observability, and performance evaluation for machine learning models, AI applications, and generative AI agents! MLflow 3.0 introduces a variety of exciting improvements to MLflow Tracking and Tracing, including:
```

> - First class model support in MLflow tracking with <APILink fn="mlflow.entities.LoggedModel">**LoggedModel**</APILink>

Reviewer: Can we describe what users can do / achieve with the feature, instead? "First class support" isn't super clear

Something like:

```
Centrally track and analyze the performance of your models and generative AI applications across all environments, from interactive queries in a development notebook through production batch or real-time serving deployments.
```

Author: Updated!

> - **Model-centric** information architecture for params and metrics in MLflow tracking

Reviewer: Can we remove all references to "model-centric" and "information architecture" in our docs? I'd rather talk about how MLflow 3.0 makes it easy to "centrally track all of the performance data for a model across environments"

Author: Got it
